File: apps/api/app/api/routes/performance.py
from decimal import Decimal
from typing import Annotated

from fastapi import APIRouter, Depends
from pydantic import BaseModel
from sqlalchemy.orm import Session

from app.api.deps import get_current_user
from app.db.session import get_db
from app.models.user import User
from app.schemas.performance import PerformanceSummaryRead
from app.services.performance import get_open_portfolio_risk, get_performance_summary

router = APIRouter(prefix="/performance", tags=["performance"])
DbSession = Annotated[Session, Depends(get_db)]
CurrentUser = Annotated[User, Depends(get_current_user)]


class OpenRiskGroupRead(BaseModel):
    group: str
    open_trade_count: int
    documented_initial_risk_amount: Decimal
    documented_initial_risk_percent: Decimal
    incomplete_risk_count: int


class ConcentrationGroupRead(BaseModel):
    group: str
    open_trade_count: int
    open_trade_percent: Decimal
    warning: bool


class AssetConcentrationRead(BaseModel):
    warning_status: str
    warning_threshold_percent: Decimal
    warnings: list[str]
    by_asset_class: list[ConcentrationGroupRead]
    by_symbol: list[ConcentrationGroupRead]
    by_sector: list[ConcentrationGroupRead]
    by_industry: list[ConcentrationGroupRead]
    review_only_notice: str


class CorrelationProxyRead(BaseModel):
    key: str
    label: str
    status: str
    open_trade_count: int
    symbols: list[str]
    message: str


class CorrelationProxySummaryRead(BaseModel):
    warning_status: str
    warnings: list[CorrelationProxyRead]
    review_only_notice: str


class OpenPortfolioRiskRead(BaseModel):
    open_trade_count: int
    complete_risk_count: int
    incomplete_risk_count: int
    documented_initial_risk_amount: Decimal
    documented_initial_risk_percent: Decimal
    max_risk_percent: Decimal
    warning_status: str
    warnings: list[str]
    asset_concentration: AssetConcentrationRead
    correlation_proxies: CorrelationProxySummaryRead
    by_strategy: list[OpenRiskGroupRead]
    by_asset_class: list[OpenRiskGroupRead]
    review_only_notice: str


class PerformanceSummaryWithOpenRiskRead(PerformanceSummaryRead):
    open_portfolio_risk: OpenPortfolioRiskRead


@router.get("/summary", response_model=PerformanceSummaryWithOpenRiskRead)
def get_summary(db: DbSession, user: CurrentUser) -> PerformanceSummaryWithOpenRiskRead:
    summary = get_performance_summary(db, user.id).model_dump()
    summary["open_portfolio_risk"] = get_open_portfolio_risk(db, user.id)
    return PerformanceSummaryWithOpenRiskRead.model_validate(summary)

File: apps/web/src/types/performance.ts
export type PerformanceByStrategy = {
  strategy_type: string;
  closed_trade_count: number;
  total_r: string;
  average_r: string | null;
  win_rate: string | null;
};

export type PerformanceByAssetClass = {
  asset_class: string;
  closed_trade_count: number;
  total_r: string;
  average_r: string | null;
  win_rate: string | null;
};

export type OpenRiskGroup = {
  group: string;
  open_trade_count: number;
  documented_initial_risk_amount: string;
  documented_initial_risk_percent: string;
  incomplete_risk_count: number;
};

export type ConcentrationGroup = {
  group: string;
  open_trade_count: number;
  open_trade_percent: string;
  warning: boolean;
};

export type AssetConcentration = {
  warning_status: "ok" | "warning";
  warning_threshold_percent: string;
  warnings: string[];
  by_asset_class: ConcentrationGroup[];
  by_symbol: ConcentrationGroup[];
  by_sector: ConcentrationGroup[];
  by_industry: ConcentrationGroup[];
  review_only_notice: string;
};

export type CorrelationProxy = {
  key: string;
  label: string;
  status: "ok" | "unknown" | "warning";
  open_trade_count: number;
  symbols: string[];
  message: string;
};

export type CorrelationProxySummary = {
  warning_status: "ok" | "unknown" | "warning";
  warnings: CorrelationProxy[];
  review_only_notice: string;
};

export type JournalScoreSummary = {
  average_entry_quality_score: string | null;
  average_stop_quality_score: string | null;
  average_exit_quality_score: string | null;
  average_discipline_score: string | null;
};

export type JournalStrategySummary = JournalScoreSummary & {
  strategy_type: string;
  reviewed_trade_count: number;
  setup_rule_followed_count: number;
  setup_rule_broken_count: number;
  setup_rule_unknown_count: number;
};

export type JournalAnalytics = JournalScoreSummary & {
  closed_trade_count: number;
  reviewed_trade_count: number;
  missing_review_count: number;
  setup_rule_followed_count: number;
  setup_rule_broken_count: number;
  setup_rule_unknown_count: number;
  min_strategy_sample_size: number;
  by_strategy: JournalStrategySummary[];
  small_sample_notice: string;
};

export type OpenPortfolioRisk = {
  open_trade_count: number;
  complete_risk_count: number;
  incomplete_risk_count: number;
  documented_initial_risk_amount: string;
  documented_initial_risk_percent: string;
  max_risk_percent: string;
  warning_status: "ok" | "unknown" | "warning";
  warnings: string[];
  asset_concentration: AssetConcentration;
  correlation_proxies: CorrelationProxySummary;
  by_strategy: OpenRiskGroup[];
  by_asset_class: OpenRiskGroup[];
  review_only_notice: string;
};

export type PerformanceSummary = {
  closed_trade_count: number;
  total_r: string;
  average_r: string | null;
  win_rate: string | null;
  best_r: string | null;
  worst_r: string | null;
  by_strategy: PerformanceByStrategy[];
  by_asset_class: PerformanceByAssetClass[];
  open_portfolio_risk: OpenPortfolioRisk;
  journal_analytics: JournalAnalytics;
};
